moz_security/sandboxes/heka/output/moz_security_tigerblood.lua

```diff
@@ -0,0 +1,198 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+# Tigerblood Output Plugin
+
+Sends tigerblood events to Mozilla Tigerblood IP reputation service using a
+hawk/JSON request.
+
+https://github.com/mozilla-services/tigerblood
+
+## Sample Configuration
+```lua
+filename = "moz_security_tigerblood.lua"
+message_matcher = "Type == 'tigerblood'"
+
+-- Configuration for pushing violations to Tigerblood
+tigerblood = {
+   base_url     = "https://tigerblood.prod.mozaws.net", -- NB: no trailing slash
+   id           = "fxa_heavy_hitters", -- hawk ID
+   key          = "hawksecret", -- hawk secret
+}
+```
+--]]
+
+local client_cfg = read_config("tigerblood") or error("no tigerblood configuration specified")
+
+local string    = require("string")
+local table     = require("table")
+local digest    = require("openssl").digest
+local hmac      = require("openssl").hmac
+local random    = require("openssl").random
+local mime      = require("mime")
+local os        = require("os")
+local http      = require("socket.http")
+local https     = require("ssl.https")
+local url       = require("socket.url")
+local ltn12     = require("ltn12")
+local cjson     = require("cjson")
+
+local function url_escape(s)
+    return s:gsub("+", "-"):gsub("/", "_"):gsub("=", "")
+end
+
+local function nonce(size)
+    return url_escape(mime.b64(random(size)))
+end
+
+local function supported_hash(algorithm)
+    return algorithm == "sha256"
+end
+
+local function hash_payload(payload, algorithm, content_type)
+    if not supported_hash(algorithm) then
+        error("unsupported hash algorithm")
+    end
+
+    local normalized = "hawk.1.payload"
+    normalized = string.format("%s\n%s", normalized, string.lower(content_type))
+    normalized = string.format("%s\n%s\n", normalized, payload)
+    return mime.b64(digest.digest(algorithm, normalized, true))
+end
+
+local function normalize_string(artifacts)
+    local normalized = {
+        "hawk.1.header",
+        artifacts.ts,
+        artifacts.nonce,
+        string.upper(artifacts.method),
+        artifacts.resource,
+        string.lower(artifacts.host),
+        artifacts.port,
+        artifacts.hash,
+        "",
+    }
+    return string.format("%s\n", table.concat(normalized, "\n"))
+end
+
+local function calculate_mac(artifacts)
+    if not supported_hash(client_cfg.algorithm) then
+        error("unsupported hash algorithm")
+    end
+    return mime.b64(hmac.hmac(client_cfg.algorithm, normalize_string(artifacts),
+        client_cfg.key, true))
+end
+
+local function hawk_artifacts(method, resource)
```

Comment:
These tables (artifacts and options) are created and GC'ed for every request, they can be constructed once and re-used.  These message should be pretty low volume so it is not critical.